// monaco-page-objects/out/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initPageObjects = exports.NotificationType = exports.Notification = exports.ScmChange = exports.ScmProvider = exports.ScmView = void 0;
const AbstractElement_1 = require("./components/AbstractElement");
const loader_1 = require("./locators/loader");
__exportStar(require("selenium-webdriver"), exports);
__exportStar(require("./locators/locators"), exports);
__exportStar(require("./components/menu/Menu"), exports);
__exportStar(require("./components/menu/MenuItem"), exports);
__exportStar(require("./components/menu/TitleBar"), exports);
__exportStar(require("./components/menu/ContextMenu"), exports);
__exportStar(require("./components/menu/WindowControls"), exports);
__exportStar(require("./components/activityBar/ActivityBar"), exports);
__exportStar(require("./components/activityBar/ViewControl"), exports);
__exportStar(require("./components/activityBar/ActionsControl"), exports);
__exportStar(require("./components/sidebar/SideBarView"), exports);
__exportStar(require("./components/sidebar/ViewTitlePart"), exports);
__exportStar(require("./components/sidebar/ViewContent"), exports);
__exportStar(require("./components/sidebar/ViewSection"), exports);
__exportStar(require("./components/sidebar/ViewItem"), exports);
__exportStar(require("./components/sidebar/WelcomeContent"), exports);
__exportStar(require("./components/sidebar/tree/default/DefaultTreeSection"), exports);
__exportStar(require("./components/sidebar/tree/default/DefaultTreeItem"), exports);
__exportStar(require("./components/sidebar/tree/custom/CustomTreeSection"), exports);
__exportStar(require("./components/sidebar/tree/custom/CustomTreeItem"), exports);
__exportStar(require("./components/sidebar/extensions/ExtensionsViewSection"), exports);
__exportStar(require("./components/sidebar/extensions/ExtensionsViewItem"), exports);
var ScmView_1 = require("./components/sidebar/scm/ScmView");
Object.defineProperty(exports, "ScmView", { enumerable: true, get: function () { return ScmView_1.ScmView; } });
Object.defineProperty(exports, "ScmProvider", { enumerable: true, get: function () { return ScmView_1.ScmProvider; } });
Object.defineProperty(exports, "ScmChange", { enumerable: true, get: function () { return ScmView_1.ScmChange; } });
__exportStar(require("./components/sidebar/scm/NewScmView"), exports);
__exportStar(require("./components/sidebar/debug/DebugView"), exports);
__exportStar(require("./components/bottomBar/BottomBarPanel"), exports);
__exportStar(require("./components/bottomBar/ProblemsView"), exports);
__exportStar(require("./components/bottomBar/Views"), exports);
__exportStar(require("./components/statusBar/StatusBar"), exports);
__exportStar(require("./components/editor/EditorView"), exports);
__exportStar(require("./components/editor/TextEditor"), exports);
__exportStar(require("./components/editor/Editor"), exports);
__exportStar(require("./components/editor/SettingsEditor"), exports);
__exportStar(require("./components/editor/DiffEditor"), exports);
__exportStar(require("./components/editor/WebView"), exports);
__exportStar(require("./components/editor/ContentAssist"), exports);
var Notification_1 = require("./components/workbench/Notification");
Object.defineProperty(exports, "Notification", { enumerable: true, get: function () { return Notification_1.Notification; } });
Object.defineProperty(exports, "NotificationType", { enumerable: true, get: function () { return Notification_1.NotificationType; } });
__exportStar(require("./components/workbench/NotificationsCenter"), exports);
__exportStar(require("./components/workbench/input/Input"), exports);
__exportStar(require("./components/workbench/input/InputBox"), exports);
__exportStar(require("./components/workbench/input/QuickOpenBox"), exports);
__exportStar(require("./components/workbench/Workbench"), exports);
__exportStar(require("./components/workbench/DebugToolbar"), exports);
__exportStar(require("./components/dialog/ModalDialog"), exports);
__exportStar(require("./conditions/WaitForAttribute"), exports);
/**
 * Initialize the page objects for your tests
 *
 * @param currentVersion version of the locators to load
 * @param baseVersion base version of the locators if you have multiple versions with diffs, otherwise leave the same as currentVersion
 * @param locatorFolder folder that contains locator files
 * @param driver WebDriver instance
 * @param browserID identifier/name of the browser (i.e. vscode)
 */
function initPageObjects(currentVersion, baseVersion, locatorFolder, driver, browserID) {
    const locators = new loader_1.LocatorLoader(currentVersion, baseVersion, locatorFolder).loadLocators();
    AbstractElement_1.AbstractElement.init(locators, driver, browserID, currentVersion);
}
exports.initPageObjects = initPageObjects;
