// pg-structure/dist/util/built-in-types.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const built_in_type_1 = __importDefault(require("../pg-structure/type/built-in-type"));
/**
 * Returns builtin types to be use in pg-structure module.
 *
 * @ignore
 * @param schema is schema name to be built-in type defined.
 * @returns builtin types
 *
 * Types are from `pg_catalog.pg_type` system table.
 */
function getBuiltinTypes(schema) {
    return [
        //
        // ────────────────────────────────────────────────────── I ──────────
        //   :::::: N U M E R I C : :  :   :    :     :        :          :
        // ────────────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "smallint", internalName: "int2" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "integer", shortName: "int", internalName: "int4" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "bigint", internalName: "int8" }),
        new built_in_type_1.default({
            category: "N",
            schema,
            oid: 0,
            classOid: 0,
            name: "decimal",
            internalName: "numeric",
            hasPrecision: true,
            hasScale: true,
        }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "numeric", hasPrecision: true, hasScale: true }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "real", internalName: "float4" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "double precision", internalName: "float8" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "smallserial", internalName: "int2" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "serial", internalName: "int4" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "bigserial", internalName: "int8" }),
        //
        // ──────────────────────────────────────────────────────── I ──────────
        //   :::::: M O N E T A R Y : :  :   :    :     :        :          :
        // ──────────────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "money" }),
        //
        // ────────────────────────────────────────────────────────── I ──────────
        //   :::::: C H A R A C T E R : :  :   :    :     :        :          :
        // ────────────────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "S", schema, oid: 0, classOid: 0, name: "character varying", shortName: "varchar", hasLength: true }),
        new built_in_type_1.default({ category: "S", schema, oid: 0, classOid: 0, name: "character", shortName: "char", hasLength: true }),
        new built_in_type_1.default({ category: "S", schema, oid: 0, classOid: 0, name: "text" }),
        //
        // ──────────────────────────────────────────────────── I ──────────
        //   :::::: B I N A R Y : :  :   :    :     :        :          :
        // ──────────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "U", schema, oid: 0, classOid: 0, name: "bytea" }),
        //
        // ────────────────────────────────────────────────────────── I ──────────
        //   :::::: D A T E   T I M E : :  :   :    :     :        :          :
        // ────────────────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "D", schema, oid: 0, classOid: 0, name: "timestamp" }),
        new built_in_type_1.default({
            category: "D",
            schema,
            oid: 0,
            classOid: 0,
            name: "timestamp without time zone",
            shortName: "timestamp",
            hasPrecision: true,
        }),
        new built_in_type_1.default({
            category: "D",
            schema,
            oid: 0,
            classOid: 0,
            name: "timestamp with time zone",
            shortName: "timestamptz",
            hasPrecision: true,
        }),
        new built_in_type_1.default({ category: "D", schema, oid: 0, classOid: 0, name: "date" }),
        new built_in_type_1.default({ category: "D", schema, oid: 0, classOid: 0, name: "time" }),
        new built_in_type_1.default({ category: "D", schema, oid: 0, classOid: 0, name: "time without time zone", shortName: "time", hasPrecision: true }),
        new built_in_type_1.default({ category: "D", schema, oid: 0, classOid: 0, name: "time with time zone", shortName: "timetz", hasPrecision: true }),
        new built_in_type_1.default({ category: "T", schema, oid: 0, classOid: 0, name: "interval", hasPrecision: true }),
        //
        // ────────────────────────────────────────────────────── I ──────────
        //   :::::: B O O L E A N : :  :   :    :     :        :          :
        // ────────────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "B", schema, oid: 0, classOid: 0, name: "boolean", shortName: "bool" }),
        //
        // ────────────────────────────────────────────────────────── I ──────────
        //   :::::: G E O M E T R I C : :  :   :    :     :        :          :
        // ────────────────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "G", schema, oid: 0, classOid: 0, name: "point" }),
        new built_in_type_1.default({ category: "G", schema, oid: 0, classOid: 0, name: "line" }),
        new built_in_type_1.default({ category: "G", schema, oid: 0, classOid: 0, name: "lseg" }),
        new built_in_type_1.default({ category: "G", schema, oid: 0, classOid: 0, name: "box" }),
        new built_in_type_1.default({ category: "G", schema, oid: 0, classOid: 0, name: "path" }),
        new built_in_type_1.default({ category: "G", schema, oid: 0, classOid: 0, name: "polygon" }),
        new built_in_type_1.default({ category: "G", schema, oid: 0, classOid: 0, name: "circle" }),
        //
        // ────────────────────────────────────────────────────────────────────── I ──────────
        //   :::::: N E T W O R K   A D D R E S S : :  :   :    :     :        :          :
        // ────────────────────────────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "I", schema, oid: 0, classOid: 0, name: "cidr" }),
        new built_in_type_1.default({ category: "I", schema, oid: 0, classOid: 0, name: "inet" }),
        new built_in_type_1.default({ category: "U", schema, oid: 0, classOid: 0, name: "macaddr" }),
        new built_in_type_1.default({ category: "U", schema, oid: 0, classOid: 0, name: "macaddr8" }),
        //
        // ──────────────────────────────────────────────────────────── I ──────────
        //   :::::: B I T   S T R I N G : :  :   :    :     :        :          :
        // ──────────────────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "V", schema, oid: 0, classOid: 0, name: "bit", hasLength: true }),
        new built_in_type_1.default({ category: "V", schema, oid: 0, classOid: 0, name: "bit varying", shortName: "varbit", hasLength: true }),
        //
        // ────────────────────────────────────────────────────────────── I ──────────
        //   :::::: T E X T   S E A R C H : :  :   :    :     :        :          :
        // ────────────────────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "U", schema, oid: 0, classOid: 0, name: "tsvector" }),
        new built_in_type_1.default({ category: "U", schema, oid: 0, classOid: 0, name: "tsquery" }),
        //
        // ──────────────────────────────────────────────── I ──────────
        //   :::::: U U I D : :  :   :    :     :        :          :
        // ──────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "U", schema, oid: 0, classOid: 0, name: "uuid" }),
        //
        // ────────────────────────────────────────────── I ──────────
        //   :::::: X M L : :  :   :    :     :        :          :
        // ────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "U", schema, oid: 0, classOid: 0, name: "xml" }),
        //
        // ──────────────────────────────────────────────── I ──────────
        //   :::::: J S O N : :  :   :    :     :        :          :
        // ──────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "U", schema, oid: 0, classOid: 0, name: "json" }),
        new built_in_type_1.default({ category: "U", schema, oid: 0, classOid: 0, name: "jsonb" }),
        new built_in_type_1.default({ category: "U", schema, oid: 0, classOid: 0, name: "jsonpath" }),
        //
        // ────────────────────────────────────────────────── I ──────────
        //   :::::: R A N G E : :  :   :    :     :        :          :
        // ────────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "R", schema, oid: 0, classOid: 0, name: "int4range" }),
        new built_in_type_1.default({ category: "R", schema, oid: 0, classOid: 0, name: "int8range" }),
        new built_in_type_1.default({ category: "R", schema, oid: 0, classOid: 0, name: "numrange" }),
        new built_in_type_1.default({ category: "R", schema, oid: 0, classOid: 0, name: "tsrange" }),
        new built_in_type_1.default({ category: "R", schema, oid: 0, classOid: 0, name: "tstzrange" }),
        new built_in_type_1.default({ category: "R", schema, oid: 0, classOid: 0, name: "daterange" }),
        //
        // ────────────────────────────────────────────────────────────────────────── I ──────────
        //   :::::: O B J E C T   I D E N T I F I E R : :  :   :    :     :        :          :
        // ────────────────────────────────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "oid" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "regproc" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "regprocedure" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "regoper" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "regoperator" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "regclass" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "regtype" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "regrole" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "regnamespace" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "regconfig" }),
        new built_in_type_1.default({ category: "N", schema, oid: 0, classOid: 0, name: "regdictionary" }),
        //
        // ────────────────────────────────────────────────── I ──────────
        //   :::::: O T H E R : :  :   :    :     :        :          :
        // ────────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "U", schema, oid: 0, classOid: 0, name: "pg_lsn" }),
        //
        // ──────────────────────────────────────────────────── I ──────────
        //   :::::: P S E U D O : :  :   :    :     :        :          :
        // ──────────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "any" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "anyelement" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "anyarray" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "anynonarray" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "anyenum" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "anyrange" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "cstring" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "internal" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "language_handler" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "fdw_handler" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "index_am_handler" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "tsm_handler" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "record" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "trigger" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "event_trigger" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "pg_ddl_command" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "void" }),
        new built_in_type_1.default({ category: "X", schema, oid: 0, classOid: 0, name: "unknown" }),
        new built_in_type_1.default({ category: "P", schema, oid: 0, classOid: 0, name: "opaque" }),
        //
        // ──────────────────────────────────────────────────── I ──────────
        //   :::::: O T H E R   S Y S T E M : :  :   :    :     :        :          :
        // ──────────────────────────────────────────────────────────────
        //
        new built_in_type_1.default({ category: "S", schema, oid: 0, classOid: 0, name: "name", internalName: "name" }),
        new built_in_type_1.default({ category: "S", schema, oid: 0, classOid: 0, name: "pg_node_tree", internalName: "pg_node_tree" }),
        new built_in_type_1.default({ category: "U", schema, oid: 0, classOid: 0, name: "aclitem", shortName: "aclitem", internalName: "aclitem" }),
        new built_in_type_1.default({ category: "U", schema, oid: 0, classOid: 0, name: "xid", shortName: "xid", internalName: "xid" }),
        new built_in_type_1.default({
            category: "A",
            schema,
            oid: 0,
            classOid: 0,
            name: "int2vector",
            shortName: "int2vector",
            internalName: "int2vector",
        }),
        new built_in_type_1.default({ category: "A", schema, oid: 0, classOid: 0, name: "oidvector", shortName: "oidvector", internalName: "oidvector" }),
        new built_in_type_1.default({
            category: "A",
            schema,
            oid: 0,
            classOid: 0,
            name: "int2vector",
            shortName: "int2vector",
            internalName: "int2vector",
        }),
        new built_in_type_1.default({
            category: "X",
            schema,
            oid: 0,
            classOid: 0,
            name: "int2vector",
            shortName: "int2vector",
            internalName: "int2vector",
        }),
        new built_in_type_1.default({ category: "D", schema, oid: 0, classOid: 0, name: "abstime", shortName: "abstime", internalName: "abstime" }),
        new built_in_type_1.default({
            category: "S",
            schema,
            oid: 0,
            classOid: 0,
            name: "pg_ndistinct",
        }),
        new built_in_type_1.default({
            category: "S",
            schema,
            oid: 0,
            classOid: 0,
            name: "pg_dependencies",
        }),
        new built_in_type_1.default({
            category: "S",
            schema,
            oid: 0,
            classOid: 0,
            name: "pg_mcv_list",
        }),
    ];
}
exports.default = getBuiltinTypes;
